gnocchi/rest/api.py

```diff
@@ -1883,6 +1890,62 @@ class BatchController(object):
     resources = ResourcesBatchController()
 
 
+class PrometheusController(rest.RestController):
+    @pecan.expose()
+    def post(self):
+        buf = snappy.uncompress(pecan.request.body)
+        f = remote_pb2.WriteRequest()
+        f.ParseFromString(buf)
+        measures = {}
+        for ts in f.timeseries:
+            attrs = dict((l.name, l.value) for l in ts.labels)
+            if ts.samples:
+                measures[attrs['__name__']] = MeasuresListSchema(
+                    [{'timestamp': s.timestamp_ms / 1000.0,
+                      'value': s.value}
+                     for s in ts.samples])
+
+        creator = pecan.request.auth_helper.get_current_user(pecan.request)
+        rid = ResourceUUID('prometheus', creator=creator)
+        r = pecan.request.indexer.get_resource('generic', rid,
+                                               with_metrics=True)
+        if not r:
+            metrics = MetricsSchema(dict((n, {}) for n in measures))
+            try:
+                r = pecan.request.indexer.create_resource(
+                    'generic', rid, creator,
+                    original_resource_id="prometheus",
+                    metrics=metrics
+                )
+            except indexer.ResourceAlreadyExists:
+                # Created in the meantime
+                need_update = True
+            else:
+                need_update = False
+        else:
+            need_update = True
+
+        if need_update:
+            if not r:
+                r = pecan.request.indexer.get_resource('generic', rid,
+                                                       with_metrics=True)
```

Comment:
I don't want to kill the mood, but looks like it can also have been deleted at this stage and then raise an exception.